### ISSUE TITLE
Optional permission

### ISSUE THREAD

**Issue by bdcomp** [fnxweb/neo-diggler#1]
Hi,

I would like to use this add-on only for navigating up through the current URL.

Is it possible to declare the pretty scary "Access your data for all websites" permission as optional [1] and add an option to some extra features (beside the navigating up)?

[1] [optional_permissions](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/optional_permissions)

**Comment by fnxweb** [comment]
I'll be honest, I'm not wholly sure what's triggering that permission.  The Mozilla site doesn't have a mapping of requested permissions to the installation warning text.

I've been assuming it's something to do with the good you have to go through to use the clipboard from the background code.

I'll have a play and see if I can figure out which it is, then I'll know what function we'll lose by making it optional.

**Comment by fnxweb** [comment]
Well, this is unpleasant;  that warning doesn't mean at _all_ what I thought (well, it sort of does, but not in the way I think either of us took it):

**./browser/locales/en-US/chrome/browser**
`/browser.properties:webextPerms.hostDescription.allUrls=Access your data for all websites`

and:

**./toolkit/components/extensions/Extension.jsm**
```
    // Format the host permissions.  If we have a wildcard for all urls,
    // a single string will suffice.  Otherwise, show domain wildcards
    // first, then individual host permissions.
    result.msgs = [];
    if (allUrls) {
      result.msgs.push(bundle.GetStringFromName("webextPerms.hostDescription.allUrls"));
    } else {
```

So, it _really_ means that the extension will be available on all sites (as opposed to, say, a known subset).  I don't think that's avoidable, and will apply to when Firefox decides to even _allow_ the extension, not when the user tries certain operations (at which point asking would make sense).  You don't want to be asked every time you visit the site ...

Unless you know better?  I'll try to have a little play either way, but since it's the URL filter, not an actual API permission, I don't think I can affect it via **optional_permissions**.

**Comment by bdcomp** [comment]
But the add-on description says:
> Access your data for all websites — injects code to the current page in order to write to the clipboard.

So, it looks like the add-on really uses this permission, not just to be enabled on all the sites.

Is it possible to make this code optional from the settings?
Or, how about to release "Neo Diggler Lite" with the only ability of navigating up through the current URL?

**Comment by fnxweb** [comment]
It's not code that's doing it, is the bit in the manifest that specifies what websites it can work in. If I didn't say “all” you'd only get the same message but for each of the allowed websites.

At least, so far as I can tell. “Access your data” send to be synonymous with “this extension works with”.  I'm not sure how *any* page-related extension can not require that permission.

I'll ask online during the week to see if my understanding is correct.

**Comment by bdcomp** [comment]
This mine very simple add-on works without this permission on any page:
[i-parcel Track and Trace](https://github.com/bdcomp/iparcelTrack)

**Comment by fnxweb** [comment]
Interesting, thanks.  Hmm, I do recall adding it after FF moaned about ... something.  I'll have a play during the week.  My other extension doesn't need it, and does _more_ so maybe I added while trying something and it can just come out now.

**Comment by fnxweb** [comment]
Can you please try 2.0.2beta1:

https://addons.mozilla.org/en-GB/firefox/addon/neo-diggler/versions/beta?page=1#version-2.0.2beta1

No changes, it just doesn't have the all-URLS permission.  Although I am getting an unlocated “missing host permissions for tab” error off the pop-up which I'll have to investigate ...  but the site-nav menu seems to work.

**Comment by bdcomp** [comment]
Yay! Version 2.0.2beta1 - navigating up works great.
Thanks.

**Comment by fnxweb** [comment]
Yeah, no promises though. The beta was the only way to prove I'd sorted out the permissions request, but now I have to make the pop up work again...

**Comment by fnxweb** [comment]
I think I've gotten it properly working now with the reduced permissions ...

https://addons.mozilla.org/en-GB/firefox/addon/neo-diggler/versions/beta?page=1#version-2.0.2beta2

**Comment by bdcomp** [comment]
Version 2.0.2beta2 - navigating up works great.

**Comment by fnxweb** [comment]
Fixed by bd3faa51b8a7ee48fc7dbbb03739159854a120ff
